src/Init/Control/Basic.lean

```diff
@@ -20,8 +20,21 @@ def Functor.discard {f : Type u → Type v} {α : Type u} [Functor f] (x : f α)
 
 export Functor (discard)
 
+/--
+An `Alternative` functor is an `Applicative` functor that has a notion of “emptyness” or “failure”
+and a binary operation `<|>` with a notion of “collecting valures” or “left-most success”.
+
+Important instances include
+* `Option`, where `failure := .none` and `<|>` returns the left-most `.some`.
+* Parser combinators typically provide an `Applicative` instance for error-handling and
+  backtracking.
```

Comment:
As highlighted in [this thread](https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/Tauto.20infinite.20loop/near/430695245), lean monads are somewhat inconsistent about whether `try catch` and/or `<|>` backtracks some or all of the state. Probably not much can be done on the docstring here other than to flag that the behavior varies and the user should consult the documentation for the relevant monad.

Reply:
How about
```suggestion
  backtracking; see the respective monad’s documentation for details.
```

Reply:
Next try:

```suggestion
  backtracking.
  
See the respective monad documentation for the precise semantics, e.g. which side effects 
are backtracked.
```

Reply:
An `Alternative` is not necessarily a monad, and the docs are often not as informative as they could be on this point.

Another way to say this would be something like:

Error recovery and state can interact subtly. For example, the implementation of `Alternative` for `OptionT (StateT \sigma Id)` keeps modifications made to the state while recovering from failure, while`StateT \sigma (OptionT Id)` discards them.

Reply:
Next try:

```suggestion
  backtracking.
  
Error recovery and state can interact subtly. For example, the implementation of `Alternative` for `OptionT (StateT σ Id)` keeps modifications made to the state while recovering from failure, while `StateT σ (OptionT Id)` discards them.
```